src/hmem_ipc_cache.c

```diff
@@ -0,0 +1,231 @@
+ * assign the mapped_addr the mapped address of the ipc_handle.
+ * Each (ipc_handle, mapped_addr) pair is stored in ofi_mr_entry.info.ipc_info as
+ * part of each mr entry.
+ * In a cache hit, the mapped_addr is retrieved from the matched mr entry. Otherwise,
+ * the mapped_addr is obtained by opening the ipc handle.
+ * 
+ * @param[in] cache the ipc cache
+ * @param[in] ipc_info the information of the ipc to be mapped.
+ * @param[out] mapped_addr the mapped address for the ipc handle.
+ * @param[out] mr_entry the matched mr_entry of the ipc_info and mapped_addr.
+ * @return int 0 on success, negative value otherwise.
+ */
+int ofi_ipc_cache_search(struct ofi_mr_cache *cache, struct smr_ipc_info *ipc_info,
+			  void **mapped_addr, struct ofi_mr_entry **mr_entry)
+{
+	struct fi_mr_attr attr;
+	struct iovec iov;
+	struct ofi_mr_entry *entry;
+	int ret;
+	ofi_genlock_lock(&cache->domain->lock);
+	iov.iov_base = (void *)ipc_info->base_address;
+	iov.iov_len = ipc_info->base_length;
+	attr.mr_iov = &iov;
+	attr.iface = ipc_info->iface;
+	attr.iov_count = 1;
+
+	ret = ofi_mr_cache_search(cache, &attr, &entry);
+	if (OFI_UNLIKELY(ret)) {
+		ofi_genlock_unlock(&cache->domain->lock);
+		return ret;
+	}
+
+	if (!entry->info.ipc_mapped_addr) {
+		/*
+		 * This indicates that the mr entry is inserted in the cache the first time 
+		 * thus does not have the ipc_mapped_addr assigned in the info.
+		 */
+		memcpy((void **)&entry->info.ipc_handle, (void **) &ipc_info->ipc_handle, IPC_HANDLE_SIZE);
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"ipc cache opens new ipc handle, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+		ret = ipc_cache_open_handle(cache, entry);
+		if (OFI_UNLIKELY(ret)) {
+			ofi_genlock_unlock(&cache->domain->lock);
+			return ret;
+		}
+		goto hit;
+	}
+	if (memcmp((void **)&entry->info.ipc_handle,
+		   ipc_info->ipc_handle, IPC_HANDLE_SIZE) == 0) {
+		/* cache hit */
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"ipc cache hit, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+		goto hit;
+	} else {
+		/* remove stale IPC handle */
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"ipc cache closes stale ipc handle, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+		cache->delete_region(cache, entry);
+		/* Update the IPC handle in the mr entry */
+		memcpy((void **)&entry->info.ipc_handle,
+		       ipc_info->ipc_handle, IPC_HANDLE_SIZE);
+		/* Open a new handle when cache misses or the ipc handle is stale */
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"ipc cache opens new ipc handle, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+		ret = ipc_cache_open_handle(cache, entry);
+		if (OFI_UNLIKELY(ret)) {
+			ofi_genlock_unlock(&cache->domain->lock);
+			return ret;
+		}
+	}
+hit:
+	*mapped_addr = entry->info.ipc_mapped_addr;
+	*mr_entry = entry;
+	ofi_genlock_unlock(&cache->domain->lock);
```

Comment:
can you add a goto out here and return ret instead of FI_SUCCESS, then add a goto on each exit condition?